File: app/src/main/java/com/nichols/dsa/concurrency/FooBar.java
package com.nichols.dsa.concurrency;

import java.util.concurrent.Semaphore;

public class FooBar {
    private int  n;
    private final Semaphore fooSemaphore;
    private final Semaphore barSemaphore;

    public FooBar(int n){
        this.n  = n;
        fooSemaphore = new Semaphore(1);
        barSemaphore = new Semaphore(0);
    }

    public void foo(Runnable printFoo) throws InterruptedException  {
        for(int i=  0; i <n; i++){
            fooSemaphore.acquire();
            printFoo.run();
            barSemaphore.release();
        }
    }
    public void bar(Runnable printBar)  throws  InterruptedException {
        for(int i  = 0; i  <  n;  i++){
            barSemaphore.acquire();
            printBar.run();
            fooSemaphore.release();
        }
    }
}

File: app/src/main/java/com/nichols/dsa/recursion/FooBar.kt
package com.nichols.dsa.recursion

fun main(){
    val g = foo(3,5)
    val f = 0
}
fun bar(x: Int, y: Int): Int {
    if(y==0) return 0

    return (x + bar(x, y-1))
}
fun foo(x: Int, y: Int): Int{
    if(y == 0)
        return 1

    return bar(x, foo(x,y-1))
}
